### ISSUE TITLE
Delegate method assumes NSURLResponse is a NSHTTPURLResponse resulting in possible crash

### ISSUE THREAD

**Issue by sgtsquiggs** [rs/SDWebImage#1109]
in [SDWebImageDownloaderOperation.m](https://github.com/rs/SDWebImage/blob/beabb1842bbde9e9c9078145e4007a17b7817739/SDWebImage/SDWebImageDownloaderOperation.m#L209) there is a mistake which will result in a crash when ```connection:didReceiveResponse:``` handles a NSURLResponse (ie, a link to a local file in cache) rather than a NSHTTPURLResponse.

The issue is a simple commutative mistake with a simple fix.

The offending delegate method:
```objc
- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response {
    
    //'304 Not Modified' is an exceptional one
    if ((![response respondsToSelector:@selector(statusCode)] || [((NSHTTPURLResponse *)response) statusCode] < 400) && [((NSHTTPURLResponse *)response) statusCode] != 304) {
        NSInteger expected = response.expectedContentLength > 0 ? (NSInteger)response.expectedContentLength : 0;
        self.expectedSize = expected;
        if (self.progressBlock) {
            self.progressBlock(0, expected);
        }

        self.imageData = [[NSMutableData alloc] initWithCapacity:expected];
        self.response = response;
        dispatch_async(dispatch_get_main_queue(), ^{
            [[NSNotificationCenter defaultCenter] postNotificationName:SDWebImageDownloadReceiveResponseNotification object:self];
        });
    }
    else {
        NSUInteger code = [((NSHTTPURLResponse *)response) statusCode];
        
        //This is the case when server returns '304 Not Modified'. It means that remote image is not changed.
        //In case of 304 we need just cancel the operation and return cached image from the cache.
        if (code == 304) {
            [self cancelInternal];
        } else {
            [self.connection cancel];
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            [[NSNotificationCenter defaultCenter] postNotificationName:SDWebImageDownloadStopNotification object:self];
        });

        if (self.completedBlock) {
            self.completedBlock(nil, nil, [NSError errorWithDomain:NSURLErrorDomain code:[((NSHTTPURLResponse *)response) statusCode] userInfo:nil], YES);
        }
        CFRunLoopStop(CFRunLoopGetCurrent());
        [self done];
    }
}
```

The offending line alone:
```objc
if ((![response respondsToSelector:@selector(statusCode)] || [((NSHTTPURLResponse *)response) statusCode] < 400) && [((NSHTTPURLResponse *)response) statusCode] != 304) {
```

The proposed fix:
```objc
if (![response respondsToSelector:@selector(statusCode)] || ([((NSHTTPURLResponse *)response) statusCode] < 400 && [((NSHTTPURLResponse *)response) statusCode] != 304)) {
```

**Comment by sgtsquiggs** [comment]
Looks like this was just fixed a few days ago by @pollarm in PR #1104. Carry on.

**Comment by lqthien** [comment]
Seems like the line is still causing crashes on 3.7.2. Our project's code runs fine if we specify 3.7.1.

**Comment by bpoplauschi** [comment]
@lqthien we will soon release 3.7.3 with this fix included (#1104).
